Closes #5371 - Introduce 'sync' ping, add 'syncUuid' to all sync-related pings

grigoryk opened the pull request:
This patch introduce an 'overarching' sync ping, which is meant to contain information
describing a sync overall, vs individual engine runs. Currently it contains global sync
errors.

syncUuid was also added to all sync-related pings, allowing us to tie together in an
analysis all pings that were emitted as part of a single "sync".


---
<!-- Text above this line will be added to the commit once "bors" merges this PR -->

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

linacambridge commented on components/support/sync-telemetry/src/main/java/mozilla/components/support/sync/telemetry/SyncTelemetry.kt:
```diff
@@ -4,28 +4,75 @@
+            syncInfo.failureReason?.let { failureReason ->
+                recordFailureReason(failureReason, Sync.failureReason)
+                Pings.sync.send()
+                return
```
I don't think this early return is right—[it's possible](https://github.com/mozilla/application-services/blob/0b392c365af1c4ce477f35f469b00288b01cf4f7/components/sync15/src/sync_multiple.rs#L326-L333) for us to sync some engines, and then get a hard error that fails the entire sync, if we get an HTTP server error, token authentication error, or other kind of network error.

So we can have some engines that succeed (and others that fail, with different reasons) and still have a global `failure_reason`.

grigoryk replied:
Ah, good call!

linacambridge commented on components/support/sync-telemetry/src/test/java/mozilla/components/support/sync/telemetry/SyncTelemetryTest.kt:
```diff
@@ -327,7 +327,7 @@ class SyncTelemetryTest {
     @Test
     fun `sends passwords telemetry pings on success`() {
-        val noGlobalError = SyncTelemetry.processPasswordsPing(SyncTelemetryPing(
+        val noGlobalError = SyncTelemetry.processLoginsPing(SyncTelemetryPing(
```
Nice catch!

Dexterp37 commented on components/support/sync-telemetry/metrics.yaml:
```diff
@@ -4,11 +4,61 @@
+    type: uuid
+    description: >
+      Unique identifier for this sync, used to correlate together individual sync pings (history, bookmarks, logins).
+    send_in_pings:
```
Yes, you can :)

After each of the listed ping is collected, then the recorded value *for that ping* is cleared. Example:

- `sync` ping is collected, then the `sync.sync_uuid` metric is cleared for the `sync` ping.
- `history_sync` ping is collected, then the `sync.sync_uuid` metric is cleared for the `history_sync` ping.
- ... and so on

grigoryk replied:
Thanks for the suggestion, much better than repeating it.

Dexterp37 commented on components/support/sync-telemetry/metrics.yaml:
```diff
@@ -4,11 +4,61 @@
+    bugs:
+      - https://github.com/mozilla-mobile/android-components/issues/5371
+    data_reviews:
+      - TODO
```
Friendly reminder to request a data review :)

Dexterp37 commented on components/support/sync-telemetry/pings.yaml:
```diff
@@ -4,6 +4,17 @@
 
+sync:
+  description: >
+    A summary sync ping, sent for every sync.
```
Since this will end up in the autogenerated docs, it might be worth clarifying a bit to give more context to people who do not know much about how sync works.

This says `sent for every sync`, but other pings say, for example,  `sent for every history sync`. Is there any way to phrase this so that disambiguates between the "engine sync" and the "global sync"?

Please ignore this if you feel like that's already clear.

grigoryk replied:
Updated the description, thanks for pointing this out.

Dexterp37 commented on components/support/sync-telemetry/src/main/java/mozilla/components/support/sync/telemetry/SyncTelemetry.kt:
```diff
@@ -4,28 +4,78 @@
+            syncInfo.engines.forEach { engineInfo ->
+                when (engineInfo.name) {
+                    "bookmarks" -> {
+                        individualBookmarksSync(syncTelemetry.uid, syncUuid, engineInfo)
```
If you change the metric definition as suggested by @linacambridge , then you will be able to set the uuid only once.

grigoryk commented:
@Dexterp37 @linacambridge finally got around to finishing this - addressed your feedback and added tests. Tests ended up being a little tricky to write, but not insurmountable.

grigoryk commented:
CI is failing because bintray seems to be having a bad day:
`[task 2020-04-02T01:42:22.078Z]             > Could not GET 'https://jcenter.bintray.com/org/jetbrains/kotlinx/kotlinx-coroutines-core/1.1.1/kotlinx-coroutines-core-1.1.1.pom'. Received status code 502 from server: Bad Gateway`

grigoryk commented:
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

How Sync implementations are behaving in the wild.

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:

Sync is one of the core features of our products; we need to know that the implementations of it that we're shipping are healthy.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

If we want to fully understand health of the sync implementation we're currently shipping, there really isn't an alternative to this approach. It maps directly to how sync operates. Any alternative that's sufficiently different wouldn't map as directly to the actual implementation (i.e., wouldn't reflect reality as well).

4) Can current instrumentation answer these questions?
Sort of; we can see how individual data type syncs are behaving, but we can't quite tell if Sync overall is functioning correctly, since the current telemetry doesn't allow us to tie together pings for different data type syncs which all run together.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki.   

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>Global sync ping, which includes a global failureReason and a global syncUuid unique to a single Sync</td>
    <td>Category 1</td>
    <td>#5371</td>
  </tr>
  <tr>
    <td>history-sync.syncUuid metric, same value as the global syncUuid</td>
    <td>Category 1</td>
    <td>#5371</td>
  </tr>
  <tr>
    <td>bookmarks-sync.syncUuid metric, same value as the global syncUuid</td>
    <td>Category 1</td>
    <td>#5371</td>
  </tr>
  <tr>
    <td>logins-sync.syncUuid metric, same value as the global syncUuid</td>
    <td>Category 1</td>
    <td>#5371</td>
  </tr>
</table>


6) How long will this data be collected?  Choose one of the following:

I want to permanently monitor this data. Grisha Kruglov + `application-services` team.

7) What populations will you measure?

All populations for all products which ship relevant a-c components.

8) If this data collection is default on, what is the opt-out mechanism for users?

Standard Glean SDK mechanism.

9) Please provide a general description of how you will analyze this data.

Redash/GLAM (when available) dashboards.

10) Where do you intend to share the results of your analysis?

Internally within sync, application-services, android-components and fenix teams.

11) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:

No.

Dexterp37 commented on components/support/sync-telemetry/metrics.yaml:
```diff
@@ -4,6 +4,47 @@
+    description: >
+      Unique identifier for this sync, used to correlate together individual pings for data types that
+      were synchronized together (history, bookmarks, logins). If a data type is synchronized by itself via
+      the legacy 'sync' API (as opposed to the Sync Manager), then 'syncUuid' will not be set on the corresponding ping.
```
```suggestion
      the legacy 'sync' API (as opposed to the Sync Manager), then this field will not be set on the corresponding ping.
```

Note that `syncUuid` is the label of the generated object in Kotlin, the field is called `sync_uuid`. We can also just not mention it by name, here.

Dexterp37 commented on components/support/sync-telemetry/src/main/java/mozilla/components/support/sync/telemetry/SyncTelemetry.kt:
```diff
@@ -21,6 +25,63 @@ const val MAX_FAILURE_REASON_LENGTH = 100
+            // We can then use 'syncUuid' to associate together all of the individual syncs that happened together.
+            // If a data type sync is executed individually via the legacy 'sync' API on specific storage layers,
+            // then the corresponding ping will not have 'syncUuid' set.
+            val globalSyncUuid = UUID.randomUUID()
+            Sync.syncUuid.set(globalSyncUuid)
```
```suggestion
            val globalSyncUuid = Sync.syncUuid.generateAndset()
```

Dexterp37 commented on components/support/sync-telemetry/src/main/java/mozilla/components/support/sync/telemetry/SyncTelemetry.kt:
```diff
@@ -156,44 +159,126 @@ object SyncTelemetry {
+    private fun individualHistorySync(hashedFxaUid: String, engineInfo: EngineInfo, syncUuid: UUID? = null) {
+        require(engineInfo.name == "history") { "Expected 'history', got ${engineInfo.name}" }
+
+        syncUuid?.let { Sync.syncUuid.set(syncUuid) }
```
I don't think this line is required: you are already setting the UUID and, because you are saying in `pings.yaml` it should go in all the pings, then it will go in all the pings.

Dexterp37 commented on components/support/sync-telemetry/src/main/java/mozilla/components/support/sync/telemetry/SyncTelemetry.kt:
```diff
@@ -156,44 +159,126 @@ object SyncTelemetry {
+    private fun individualBookmarksSync(hashedFxaUid: String, engineInfo: EngineInfo, syncUuid: UUID? = null) {
+        require(engineInfo.name == "bookmarks") { "Expected 'bookmarks', got ${engineInfo.name}" }
+
+        syncUuid?.let { Sync.syncUuid.set(syncUuid) }
```
I don't think this line is required: you are already setting the UUID and, because you are saying in `pings.yaml` it should go in all the pings, then it will go in all the pings.

Dexterp37 commented on components/support/sync-telemetry/src/main/java/mozilla/components/support/sync/telemetry/SyncTelemetry.kt:
```diff
@@ -156,44 +159,126 @@ object SyncTelemetry {
+    private fun individualLoginsSync(hashedFxaUid: String, engineInfo: EngineInfo, syncUuid: UUID? = null) {
+        require(engineInfo.name == "passwords") { "Expected 'passwords', got ${engineInfo.name}" }
+
+        syncUuid?.let { Sync.syncUuid.set(syncUuid) }
```
I don't think this line is required: you are already setting the UUID and, because you are saying in `pings.yaml` it should go in all the pings, then it will go in all the pings.

Dexterp37 commented on components/support/sync-telemetry/src/main/java/mozilla/components/support/sync/telemetry/SyncTelemetry.kt:
```diff
@@ -21,6 +25,63 @@ const val MAX_FAILURE_REASON_LENGTH = 100
+            syncInfo.engines.forEach { engineInfo ->
+                when (engineInfo.name) {
+                    "bookmarks" -> {
+                        individualBookmarksSync(syncTelemetry.uid, engineInfo, globalSyncUuid)
```
I don't think you need to pass `globalSyncUuid` around, see my comments below.

grigoryk replied:
Thanks, that makes sense. I guess I was concerned that the value would be cleared after ping submission, but I'm guessing it's cleared for a specific ping that's being submitted, and not for others?
Tests seem to pass with that change.

Dexterp37 replied:
Yes, values are stored in ping-specific stores. This is the big feature that allows easy configuration of new custom ping and adding new metrics to existing ping.

Our docs somethimes _can_ be confusing! Please let us know if they are in this case!

liuche commented:
@grigoryk could you be just a little more specific in the review request form, "How Sync implementations are behaving in the wild." is pretty vague. I think there are good explanations sprinkled across the original issue, and in docs, but this form is meant to be a single place to look quickly and understand what this data collection is doing.

grigoryk commented:
@Dexterp37 thanks for feedback, addressed your comments.

grigoryk commented:
@liuche I've updated the data review form based on https://github.com/mozilla-mobile/android-components/pull/3092. Let me know if this is clearer, or if you'd like to see more detail.

liuche commented:
@boek has a data review in flight, I was just doing some drive-by commenting :P

Dexterp37 commented on components/support/sync-telemetry/pings.yaml:
```diff
@@ -4,6 +4,23 @@
+  notification_emails:
+    - sync-core@mozilla.com
+  data_reviews:
+    - TODO
```
Just a reminder to update the link to the data review **response** after you have one. Here and in the other yaml files.

grigoryk replied:
Yup, waiting for that, hence a `TODO` :)

grigoryk commented:
bors r=csadilek,Dexterp37,boek
